### PR TITLE
[Powershell BYOK] Added BYOK Powershell commandlets as per spec

### PR DESCRIPTION
**Pull request by adibalwani**
- This is the initial review for the feature with all docs as LoremIpsum
- Need to wait for an official build from C# SDK in-order to build

### CODE REVIEW

File: src/Common/Common.Api/Admin/AdminClient.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+            return this.Client.Admin.AddPowerBIEncryptionKey(tenantKeyCreationRequest);
+        }
+
+        public IEnumerable<TenantKey> GetPowerBIEncryptionKeys()
+        {
+            return this.Client.Admin.GetPowerBIEncryptionKeys()?.Value;
+        }
+
+        public IEnumerable<Dataset> GetPowerBIWorkspaceEncryptionStatus(string groupId)
```
Comment (CodeCyclone) [inline src/Common/Common.Api/Admin/AdminClient.cs]: Do we want to return the entire dataset or just a property of the dataset?

File: src/Common/Common.Api/Common.Api.csproj [hunk 2]
```diff
@@ -8,8 +8,8 @@
   </PropertyGroup>
   
   <ItemGroup>
-    <PackageReference Include="Microsoft.PowerBI.Api" Version="2.0.121817003" />
-    <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
+    <PackageReference Include="Microsoft.PowerBI.Api" Version="2.5.1905001" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.1" />
```
Comment (CodeCyclone) [inline src/Common/Common.Api/Common.Api.csproj]: This will be interesting as Newtonsoft.Json causes us issues in the past as PowerShell sessions can only load one version of Newtonsoft.Json which is used by many other PowerShell cmdlets. This could also impact Cloud Shell so we could try to roll out but it may have to be backed out. Is there a reason why the version was upgraded?

File: src/Modules/Admin/Commands.Admin.Test/AddPowerBIEncryptionKeyTests.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+        [TestMethod]
+        [TestCategory("Interactive")]
+        [TestCategory("SkipWhenLiveUnitTesting")] // Ignore for Live Unit Testing
+        public void EndToEndAddPowerBIEncryptionKey()
+        {
+            using (var ps = System.Management.Automation.PowerShell.Create())
+            {
+                // Arrange
+                ProfileTestUtilities.ConnectToPowerBI(ps, PowerBIEnvironmentType.OneBox);
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin.Test/AddPowerBIEncryptionKeyTests.cs]: Remove the PowerBIEnvironmentType input, it won't compile on Release builds as OneBox is conditionalized out of the compile

File: src/Modules/Admin/Commands.Admin.Test/GetPowerBIEncryptionKeysTests.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        [TestMethod]
+        [TestCategory("Interactive")]
+        [TestCategory("SkipWhenLiveUnitTesting")] // Ignore for Live Unit Testing
+        public void EndToEndGetPowerBIEncryptionKeys()
+        {
+            using (var ps = System.Management.Automation.PowerShell.Create())
+            {
+                // Arrange
+                ProfileTestUtilities.ConnectToPowerBI(ps, PowerBIEnvironmentType.OneBox);
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin.Test/GetPowerBIEncryptionKeysTests.cs]: Remove the PowerBIEnvironmentType input

File: src/Modules/Admin/Commands.Admin.Test/GetPowerBIWorkspaceEncryptionStatusTests.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+        [TestMethod]
+        [TestCategory("Interactive")]
+        [TestCategory("SkipWhenLiveUnitTesting")] // Ignore for Live Unit Testing
+        public void EndToEndGetPowerBIWorkspaceEncryptionStatus()
+        {
+            using (var ps = System.Management.Automation.PowerShell.Create())
+            {
+                // Arrange
+                ProfileTestUtilities.ConnectToPowerBI(ps, PowerBIEnvironmentType.EDog);
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin.Test/GetPowerBIWorkspaceEncryptionStatusTests.cs]: Remove the EDog input as it won't compile in Release

File: src/Modules/Admin/Commands.Admin.Test/RotatePowerBIEncryptionKeyTests.cs [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+        [TestMethod]
+        [TestCategory("Interactive")]
+        [TestCategory("SkipWhenLiveUnitTesting")] // Ignore for Live Unit Testing
+        public void EndToEndRotatePowerBIEncryptionKey()
+        {
+            using (var ps = System.Management.Automation.PowerShell.Create())
+            {
+                // Arrange
+                ProfileTestUtilities.ConnectToPowerBI(ps, PowerBIEnvironmentType.OneBox);
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin.Test/RotatePowerBIEncryptionKeyTests.cs]: Remove OneBox input

File: src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        #region Parameters
+
+        [Parameter(Mandatory = true)]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true)]
+        public string KeyVaultKeyUri { get; set; }
+
+        [Parameter(Mandatory = false)]
+        public bool Default { get; set; } = false;
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Use switches instead of booleans

File: src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        [Parameter(Mandatory = false)]
+        public bool Activate { get; set; } = false;
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Same here, use switches instead of booleans

Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Use three Parameter Sets, one for Default and the other Activate and one for both, set the default as the default parameter set. That way they have to declare at least one of the switches and they don't have to pass true\false, they just -SwitchName

File: src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            using (var client = this.CreateClient())
+            {
+                try
+                {
+                    var response = client.Admin.AddPowerBIEncryptionKey(Name, KeyVaultKeyUri, Default, Activate);
+                    this.Logger.WriteObject(response);
+                }
+                catch (Exception ex)
+                {
+                    this.Logger.ThrowTerminatingError(ex);
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: One issue with this is any error is terminating for PowerShell. Typically you just let the error be written and then it is up to the user to decide if the error is terminating or not by setting -ErrorAction Stop. If you didn't have the try catch it would be terminating too just because the error bubbles out but when you do throw terminating then it will get wrapped

File: src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                }
+            }
+        }
+
+        private void ThrowNotSupportedException(string parameterName, object parameterValue)
+        {
+            var notSupportedException = new NotSupportedException($"{parameterName} value {parameterValue} is not yet supported");
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Remove the word 'yet'

Comment (CodeCyclone) [reply src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Might want to reword this to: "Parameter {parameterName} with value {parameterValue} is not supported"

File: src/Modules/Admin/Commands.Admin/GetPowerBIEncryptionKeys.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+using System;
+using System.Management.Automation;
+using Microsoft.PowerBI.Common.Client;
+
+namespace Microsoft.PowerBI.Commands.Admin
+{
+    [Cmdlet(CmdletVerb, CmdletName)]
+    public class GetPowerBIEncryptionKeys : PowerBIClientCmdlet
+    {
+        public const string CmdletName = "PowerBIEncryptionKeys";
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/GetPowerBIEncryptionKeys.cs]: Drop the 's', cmdlet names shouldn't be plural so it should be "PowerBIEncryptionKey"

File: src/Modules/Admin/Commands.Admin/GetPowerBIEncryptionKeys.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        public override void ExecuteCmdlet()
+        {
+            using (var client = this.CreateClient())
+            {
+                try
+                {
+                    var response = client.Admin.GetPowerBIEncryptionKeys();
+                    this.Logger.WriteObject(response, true);
+                }
+                catch (Exception ex)
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/GetPowerBIEncryptionKeys.cs]: Same comment as other cmdlet

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+        public GetPowerBIWorkspaceEncryptionStatus() : base() { }
+
+        public GetPowerBIWorkspaceEncryptionStatus(IPowerBIClientCmdletInitFactory init) : base(init) { }
+
+        #region Parameters
+
+        [Parameter(Mandatory = true)]
+        public string Name { get; set; }
+
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: Should we support ID too?

Comment (CodeCyclone) [reply src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: May want to support a pipeline too where you can pass in a workspace

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+            {
+                try
+                {
+                    var workspaces = this.GetWorkspaces(client);
+                    var matchedWorkspace = this.GetMatchingWorkspace(workspaces);
+                    var response = client.Admin.GetPowerBIWorkspaceEncryptionStatus(matchedWorkspace.Id.ToString());
+
+                    this.Logger.WriteObject(response);
+                }
+                catch (Exception ex)
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: Same comment as other cmdlets

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+            }
+        }
+
+        private IEnumerable<Workspace> GetWorkspaces(IPowerBIApiClient client)
+        {
+            string nameFilter = $"name eq '{this.Name}'";
+            var workspaces = client.Workspaces.GetWorkspacesAsAdmin(default, nameFilter, 1, default);
+            if (workspaces == null)
+            {
+                throw new Exception("No workspaces are found");
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: Drop the 's' for workspaces.
You might just want to write an error to the logger instead of throwing an exception. It could be a terminating error too

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        #endregion
+
+        public override void ExecuteCmdlet()
+        {
+            using (var client = this.CreateClient())
+            {
+                try
+                {
+                    var workspaces = this.GetWorkspaces(client);
+                    var matchedWorkspace = this.GetMatchingWorkspace(workspaces);
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: It looks like GetWorkspaces does a filter match, why is GetMatchingWorkspaces also needed as workspaces have to have unique names?

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+            var workspaces = client.Workspaces.GetWorkspacesAsAdmin(default, nameFilter, 1, default);
+            if (workspaces == null)
+            {
+                throw new Exception("No workspaces are found");
+            }
+
+            return workspaces;
+        }
+
+        private Workspace GetMatchingWorkspace(IEnumerable<Workspace> workspaces)
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: Not sure why this method is needed

File: src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+using Microsoft.PowerBI.Common.Api;
+using Microsoft.PowerBI.Common.Client;
+
+namespace Microsoft.PowerBI.Commands.Admin
+{
+    [Cmdlet(CmdletVerb, CmdletName)]
+    public class RotatePowerBIEncryptionKey : PowerBIClientCmdlet
+    {
+        public const string CmdletName = "PowerBIEncryptionKey";
+        public const string CmdletVerb = "Rotate";
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs]: Only use approved verbs. You can use rotate as an alias for the cmdlet.
https://docs.microsoft.com/en-us/powershell/developer/cmdlet/approved-verbs-for-windows-powershell-commands 
I recommend either 'Set', 'Mount', 'Update', 'Deploy', 'Register' but I think Set is the best one

File: src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+        public RotatePowerBIEncryptionKey() : base() { }
+
+        public RotatePowerBIEncryptionKey(IPowerBIClientCmdletInitFactory init) : base(init) { }
+
+        #region Parameters
+
+        [Parameter(Mandatory = true)]
+        public string Name { get; set; }
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs]: You may want to check null or empty on the strings

Comment (CodeCyclone) [reply src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs]: Same for other cmdlets

File: src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+                }
+            }
+        }
+
+        private IEnumerable<TenantKey> GetEncryptionKeys(IPowerBIApiClient client)
+        {
+            var tenantKeys = client.Admin.GetPowerBIEncryptionKeys();
+            if (tenantKeys == null)
+            {
+                throw new Exception("No encryption keys is set");
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs]: is -> are

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+external help file: Microsoft.PowerBI.Commands.Admin.dll-Help.xml
+Module Name: MicrosoftPowerBIMgmt.Admin
+online version:
+schema: 2.0.0
+---
+
+# Add-PowerBIEncryptionKey
+
+## SYNOPSIS
+{{Fill in the Synopsis}}
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Fill in

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+
+## SYNTAX
+
+```
+Add-PowerBIEncryptionKey -Name <String> -KeyVaultKeyUri <String> [-Default <Boolean>] [-Activate <Boolean>]
+ [<CommonParameters>]
+```
+
+## DESCRIPTION
+{{Fill in the Description}}
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Fill in

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\> {{ Add example code here }}
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Provide examples

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+```
+
+{{ Add example description here }}
+
+## PARAMETERS
+
+### -Activate
+{{Fill Activate Description}}
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Fill in, repeat for rest of file where there is {{XXXX}}

File: src/Modules/Admin/Commands.Admin/help/Get-PowerBIEncryptionKeys.md [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+external help file: Microsoft.PowerBI.Commands.Admin.dll-Help.xml
+Module Name: MicrosoftPowerBIMgmt.Admin
+online version:
+schema: 2.0.0
+---
+
+# Get-PowerBIEncryptionKeys
+
+## SYNOPSIS
+{{Fill in the Synopsis}}
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Get-PowerBIEncryptionKeys.md]: Fill in file where {{XXXX}} are

File: src/Modules/Admin/Commands.Admin/help/MicrosoftPowerBIMgmt.Admin.md [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+Module Name: MicrosoftPowerBIMgmt.Admin
+Module Guid: bbf8be8e-0dcd-4cd8-b868-63e96371a47f
+Download Help Link: {{Please enter FwLink manually}}
+Help Version: {{Please enter version of help manually (X.X.X.X) format}}
+Locale: en-US
+---
+
+# MicrosoftPowerBIMgmt.Admin Module
+## Description
+{{Manually Enter Description Here}}
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/MicrosoftPowerBIMgmt.Admin.md]: Fill in, the fields will get auto-filled when you fill out the other files

File: src/Common/Common.Api/Admin/AdminClient.cs (hunk 1 above)
Comment (adibalwani) [reply src/Common/Common.Api/Admin/AdminClient.cs]: Entire Dataset

File: src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs (hunk 19 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs]: Powershell cmdlet by default checks for empty string and returns errors

File: src/Common/Common.Api/Common.Api.csproj (hunk 2 above)
Comment (adibalwani) [reply src/Common/Common.Api/Common.Api.csproj]: Removed this dependecny and suppressed the warning

File: src/Modules/Admin/Commands.Admin.Test/AddPowerBIEncryptionKeyTests.cs (hunk 3 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin.Test/AddPowerBIEncryptionKeyTests.cs]: Done

File: src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs (hunk 9 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Not catching the exception anymore

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs (hunk 13 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: It is a good suggestion but is not part of V1 private preview. I'll check with pm if he needs this before going public

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs (hunk 15 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: Logging a termination exception

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs (hunk 16 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: It is not needed anymore. Updated the code

File: src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs (hunk 17 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/GetPowerBIWorkspaceEncryptionStatus.cs]: Removed this method

File: src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs (hunk 18 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/RotatePowerBIEncryptionKey.cs]: Using set and added alias Rotate

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md (hunk 21 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Temporarily filled in with the docs from spec. Will update this before going public

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md (hunk 24 above)
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Will get the exact strings from pm and update it for public preview. Temporarily filled in with the spec strings

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+external help file: Microsoft.PowerBI.Commands.Admin.dll-Help.xml
+Module Name: MicrosoftPowerBIMgmt.Admin
+online version:
+schema: 2.0.0
+---
+
+# Add-PowerBIEncryptionKey
+
+## SYNOPSIS
+Adds an encryption key
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: For Power BI dataset

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+Add-PowerBIEncryptionKey -Name <String> -KeyVaultKeyUri <String> [-Default] [-Activate] [<CommonParameters>]
+```
+
+### Activate
+```
+Add-PowerBIEncryptionKey -Name <String> -KeyVaultKeyUri <String> [-Activate] [<CommonParameters>]
+```
+
+## DESCRIPTION
+The commandlet will add an encryption key.
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Remove cmdlet name

Comment (CodeCyclone) [reply src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Look at examples of other doc for reference, for example https://docs.microsoft.com/en-us/powershell/module/microsoftpowerbimgmt.reports/Get-PowerBIReport?view=powerbi-ps

Comment (CodeCyclone) [reply src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: State what permissions is needed to invoke

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+
+### Example 1
+```
+PS C:\> Add-PowerBIEncryptionKey -Name 'testName' -KeyVaultKeyUri 'Uri' -Default -Activate
+```
+
+## PARAMETERS
+
+### -Activate
+If the Activate flag is $true, all of the capacities that don't have activated the BYOK functionality will immediately use this key
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: For switches, look at doc for other cmdlets that have switch such as -AsString https://docs.microsoft.com/en-us/powershell/module/microsoftpowerbimgmt.profile/get-powerbiaccesstoken?view=powerbi-ps

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+Required: False
+Position: Named
+Default value: False
+Accept pipeline input: False
+Accept wildcard characters: False
+```
+
+### -KeyVaultKeyUri
+The uri of the KeyVaultKeyUri parameter is an Uri to the version of the AzureKeyVault key to be used.
+The keyvault key need to adhere to the requirements and recommendations that SQL Azure has for BYOK
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Don't abbreviate BYOK

File: src/Modules/Admin/Commands.Admin/help/MicrosoftPowerBIMgmt.Admin.md [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/MicrosoftPowerBIMgmt.Admin.md]: Add something here, example https://docs.microsoft.com/en-us/powershell/module/microsoftpowerbimgmt.data/?view=powerbi-ps

File: src/Directory.Build.props [hunk 32]
```diff
@@ -16,6 +16,7 @@ Licensed under the MIT License.
     <PowerShellExe Condition="'$(PowerShellExe)' == ''">%WINDIR%\System32\WindowsPowerShell\v1.0\powershell.exe</PowerShellExe>
     <PowerShellCommonArgs>-NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted</PowerShellCommonArgs>
     <LangName Condition="'$(LangName)' == ''">en-US</LangName>
+    <NoWarn>NU1605</NoWarn>
```
[the remark this answers is not in the document]
Comment (adibalwani) [reply src/Directory.Build.props]: Added comment and updated NewtonSoft.json since test were failing using older version

File: src/Modules/Admin/Commands.Admin/help/MicrosoftPowerBIMgmt.Admin.md [hunk 33]
```diff
@@ -0,0 +1,25 @@
...
+Module Name: MicrosoftPowerBIMgmt.Admin
+Module Guid: bbf8be8e-0dcd-4cd8-b868-63e96371a47f
+Download Help Link: {{Please enter FwLink manually}}
+Help Version: {{Please enter version of help manually (X.X.X.X) format}}
+Locale: en-US
+---
+
+# MicrosoftPowerBIMgmt.Admin Module
+## Description
+PowerShell cmdlets for managing and interacting with Power BI admin services.
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/MicrosoftPowerBIMgmt.Admin.md]: admin -> administrative
I would word this slightly different: PowerShell cmdlets for performing administrative duties the across Power BI service.

File: src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+external help file: Microsoft.PowerBI.Commands.Admin.dll-Help.xml
+Module Name: MicrosoftPowerBIMgmt.Admin
+online version:
+schema: 2.0.0
+---
+
+# Set-PowerBIEncryptionKey
+
+## SYNOPSIS
+Rotate the encryption key
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md]: Mention product. Mention Set here instead of rotate and in the description use Rotates

File: src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md [hunk 35]
```diff
@@ -0,0 +1,75 @@
...
+
+## SYNTAX
+
+```
+Set-PowerBIEncryptionKey -Name <String> -KeyVaultKeyUri <String> [<CommonParameters>]
+```
+
+## DESCRIPTION
+The commandlet will rotate the customer owned key. Before you run this command, make sure you log in using Connect-PowerBIServiceAccount.
+Requires TenantReadWriteAll permission
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md]: Same comment as other doc for scopes

File: src/Modules/Admin/Commands.Admin/help/Get-PowerBIWorkspaceEncryptionStatus.md [hunk 36]
```diff
@@ -0,0 +1,60 @@
...
+
+## SYNTAX
+
+```
+Get-PowerBIWorkspaceEncryptionStatus -Name <String> [<CommonParameters>]
+```
+
+## DESCRIPTION
+Retrives the encryption status of dataset for a given workspace name. Before you run this command, make sure you log in using Connect-PowerBIServiceAccount.
+Requires TenantReadAll or TenantReadWriteAll permission
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Get-PowerBIWorkspaceEncryptionStatus.md]: Same comment as other doc for scopes

File: src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md [hunk 37]
```diff
@@ -0,0 +1,75 @@
...
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\> Set-PowerBIEncryptionKey -Name 'testName' -KeyVaultKeyUri 'Uri'
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md]: Improve example, use fake name and show what KeyVaultUri looks like

File: src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md [hunk 38]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+## PARAMETERS
+
+### -KeyVaultKeyUri
+The uri of the KeyVaultKeyUri parameter is an Uri to the version of the AzureKeyVault key to be used.
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/help/Set-PowerBIEncryptionKey.md]: Don't mention the parameter name. AzureKeyVault to "Azure Key Vault"

File: src/Common/Common.Api/Common.Api.csproj [hunk 39]
```diff
@@ -8,8 +8,7 @@
   </PropertyGroup>
   
   <ItemGroup>
-    <PackageReference Include="Microsoft.PowerBI.Api" Version="2.0.121817003" />
-    <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
```
[the remark this answers is not in the document]
Comment (adibalwani) [reply src/Common/Common.Api/Common.Api.csproj]: It is not being used anywhere in the code. We don't need it. Not sure why it was here in the first place

File: src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs [hunk 40]
```diff
@@ -0,0 +1,77 @@
...
+
+        [Parameter(ParameterSetName = ActivateParameterSet, Mandatory = false)]
+        [Parameter(ParameterSetName = DefaultAndActivateParameterSet, Mandatory = false)]
+        public SwitchParameter Activate { get; set; }
+
+        #endregion
+
+        protected override void BeginProcessing()
+        {
+            this.Logger.WriteWarning("This cmdlet is in private preview and may not work for your tenant");
```
[the remark this answers is not in the document]
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Updated the text and move it to a constants file

File: src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs [hunk 41]
```diff
@@ -0,0 +1,77 @@
...
+            base.BeginProcessing();
+
+            if (this.Default == false)
```
[the remark this answers is not in the document]
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/AddPowerBIEncryptionKey.cs]: Allowing -Default or -Activate or -default and -activate now

File: src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md [hunk 42]
```diff
@@ -0,0 +1,117 @@
...
+external help file: Microsoft.PowerBI.Commands.Admin.dll-Help.xml
+Module Name: MicrosoftPowerBIMgmt.Admin
+online version:
+schema: 2.0.0
+---
+
+# Add-PowerBIEncryptionKey
+
+## SYNOPSIS
+Adds an encryption key for powerbi workspaces in Capacity
```
[the remark this answers is not in the document]
Comment (adibalwani) [reply src/Modules/Admin/Commands.Admin/help/Add-PowerBIEncryptionKey.md]: Made changes as per suggestion

File: src/Modules/Admin/Commands.Admin.Test/RotatePowerBIEncryptionKeyTests.cs [hunk 43]
```diff
@@ -0,0 +1,174 @@
...
+using Microsoft.PowerBI.Commands.Common.Test;
+using Microsoft.PowerBI.Commands.Profile.Test;
+using Microsoft.PowerBI.Common.Api;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Microsoft.PowerBI.Commands.Admin.Test
+{
+    [TestClass]
+    public class RotatePowerBIEncryptionKeyTests
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin.Test/RotatePowerBIEncryptionKeyTests.cs]: Change name of tests too but its minor

File: src/Modules/Admin/Commands.Admin/Constants.cs [hunk 44]
```diff
@@ -0,0 +1,12 @@
+﻿/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+namespace Microsoft.PowerBI.Commands.Admin
+{
+    public static class Constants
+    {
+        public const string PRIVATE_PREVIEW_WARNING = "This cmdlet is in private preview and may not work without activation in your tenant";
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/Constants.cs]: Minor: don't name in all caps

File: src/Modules/Admin/Commands.Admin/MicrosoftPowerBIMgmt.Admin.psd1 [hunk 45]
```diff
@@ -0,0 +1,123 @@
...
+Author = 'Microsoft Corporation'
+
+# Company or vendor of this module
+CompanyName = 'Microsoft Corporation'
+
+# Copyright statement for this module
+Copyright = 'Microsoft Corporation. All rights reserved.'
+
+# Description of the functionality provided by this module
+Description = 'Microsoft Power BI PowerShell - Admin cmdlets for Microsoft Power BI'
```
Comment (CodeCyclone) [inline src/Modules/Admin/Commands.Admin/MicrosoftPowerBIMgmt.Admin.psd1]: Admin -> Administration
